### DISCUSSION TITLE
Debug nvim crash with plugins

### DISCUSSION THREAD

**Discussion by davidcortesortuno** [LazyVim/LazyVim#1322] [Q&A]
Hi, I recently updated LazyVim and it started crashing after I typed some characters. So I removed all nvim config files and cache, re-installed LazyVim from scratch and when I open neovim it crashes immediately. If I use `nvim --noplugin` it works fine. 
I wanted to ask what's the best way to debug or find where the error is. I looked into some `log` files but did not find much information.
I'm using neovim 0.9.1 in Opensuse Tumbleweed
I also tried neovim nightly's appimage but it crashes immediately as well

This is the info from `nvim -v`:

```
NVIM v0.9.1
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc  -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wvla -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -fno-common -Wno-unused-result -Wimplicit-fallthrough -fdiagnostics-color=auto -fstack-protector-strong -DUNIT_TESTING -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/usr/include/luajit-5_1-2.1 -I/usr/include -I/usr/include/lua5.1 -I/home/abuild/rpmbuild/BUILD/neovim-0.9.1/build/src/nvim/auto -I/home/abuild/rpmbuild/BUILD/neovim-0.9.1/build/include -I/home/abuild/rpmbuild/BUILD/neovim-0.9.1/build/cmake.config -I/home/abuild/rpmbuild/BUILD/neovim-0.9.1/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Comment by dpetka2001** [comment]
Disable your plugins and start enabling them one by one. If you use version control to back them up, just revert to a previous state where you know for sure it didn't crash and start working from there regarding the changes that happened after.

**Reply by davidcortesortuno** [reply]
I'm only using LazyVim's default plugins. How can I disable them one by one?

**Reply by dpetka2001** [reply]
I couldn't reproduce any crash with default Lazyvim starter template. Can you test the following?
```shell
1. mkdir ~/.config/lazytest
2. git clone https://github.com/LazyVim/starter ~/.config/lazytest
3. NVIM_APPNAME=lazytest nvim
4. wait for the plugins to install
5. After it's finished restart Neovim with the same command in step 3
```
Do you experience any crashes when you open Neovim after the previous procedure?

**Reply by davidcortesortuno** [reply]
Hi, yes, it crashes instantly again. The plugins were installed correctly though.

**Reply by dpetka2001** [reply]
I'm not able to reproduce this on my system. Maybe something to do with your Neovim version? Are you able to run `:checkhealth` once inside Neovim? Also try uninstalling and re-installing Neovim again. Also check in `~/.local/state/nvim/` and `~/.cache/nvim/` for log messages and see if they have any errors inside which can help you decipher what is going on.

**Reply by dpetka2001** [reply]
Does it crash the instance when you're opening Neovim or are you doing some specific action and it crashes?

**Reply by davidcortesortuno** [reply]
It crashes the instance just when I open nvim from the terminal. I tried different terminals. It does not crash if I do not use LazyVim

**Reply by amaanq** [reply]
that provides nothing helpful at all, a gdb backtrace at the minimum would be useful - compile neovim with debug syms

**Reply by davidcortesortuno** [reply]
I know, sorry, that's why in my question I was asking how to debug or get more information from nvim about the problem, as I could not find anything in the log files :/

**Comment by abeldekat** [comment]
Hello @davidcortesortuno 

In addition to the excellent comments from @dpetka2001  and @amaanq , I would like to suggest an approach regarding this question:

> I'm only using LazyVim's default plugins. How can I disable them one by one?

The core modules, defining the plugins per category, can be enabled incrementally. Note that I do not regard this approach to be a best practice, but it worked for me.

Start by saving the snippet below as `minimal_init.lua` in for example ~/.config/nvim:

```lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath })
end
vim.opt.rtp:prepend(lazypath)

-- activate modules incrementally
local plugins_step1 = {
  "folke/tokyonight.nvim",
  { "LazyVim/LazyVim", import = "lazyvim.plugins.core" },
  -- We now have lazy.nvim, LazyVim and tokyonight
  -- Start uncommenting the lines below, one at a time
  --
  -- { import = "lazyvim.plugins.util" },
  -- { import = "lazyvim.plugins.treesitter" },
  -- { import = "lazyvim.plugins.coding" },
  -- { import = "lazyvim.plugins.colorscheme" },
  -- { import = "lazyvim.plugins.editor" },
  -- { import = "lazyvim.plugins.ui" },
}

-- lsp has an init.lua and cannot be tested in isolation using import
local plugins_step2 = {
  "folke/tokyonight.nvim",
  { "LazyVim/LazyVim", import = "lazyvim.plugins" }, -- now including lsp
}

-- START:
-- test incrementally: uncomment lines in variable plugins_step1, restart on each activated line
local plugins = plugins_step1
-- PROCEED:
-- uncomment the line below when all modules in plugins_step1 succeed
-- local plugins = plugins_step2

require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd.colorscheme("tokyonight")
-- add anything else here
```

Now, in you nvim folder, run `nvim -u minimal_init.lua minimal_init.lua`
Start testing incrementally by uncommenting lines in variable `plugins_step1`.
When you activate a line, restart nvim: `nvim -u minimal_init.lua minimal_init.lua`
Finish by uncommenting the line containing `plugins_step2`

**Reply by davidcortesortuno** [reply]
Thanks :) Following your instructions, `nvim` crashes only when I uncomment `{ import = "lazyvim.plugins.ui" }`, especially for long files. Does it have a log file somewhere?

**Reply by abeldekat** [reply]
That's very surprising!
The plugins in UI are listed [here](https://www.lazyvim.org/plugins/ui)

> It crashes the instance just when I open nvim from the terminal. I tried different terminals. It does not crash if I do not use LazyVim

Given your previous comment, the new situation seems to be:
> especially for long files.

Some questions:
- Does this mean that regular files crash only sometimes? 
- Does it matter which type of file you open?
- What happens when neovim crashes. Does it freeze, is there any feedback?

My first step would be to disable noice. Uncomment all lines in `plugins_step1`, including the line containing `lazyvim.plugins.ui`. 
After that line, add:
> { "folke/noice.nvim", enabled = false },

You can find the logs for this minimal_init in `~/.config/nvim/.repro/state/nvim`

**Reply by davidcortesortuno** [reply]
Ok, so it crashes for any file, even short files. I tried Python, Lua and LaTeX files. By crashing, I mean the program exists immediately without any feedback, that's why I opened this issue. It also happens with `nvim`'s nightly build.
I disabled `noice` as you suggested and `nvim` works fine now  :O

**Reply by abeldekat** [reply]
> I disabled noice as you suggested and nvim works fine now

That's great!

**Reply by dpetka2001** [reply]
The noice log is under `~/.local/state/nvim/noice.log`. If there is not any error that you can understand as to what is wrong in there, you should open an issue at `noice` repo along with the log file and when folke comes back at end of month, he might be able to figure out what is wrong.

Edit: @abeldekat excellent analysis. I wouldn't have thought of that this way. I would go around disabling all plugins and enabling one by one. Nice to learn new ways of doing things from you as always :smile:

**Reply by davidcortesortuno** [reply]
I can't find any `noice.log` file, is it necessary to add an option to generate the log file?

**Reply by dpetka2001** [reply]
I believe that is because you disabled `noice`? Just enable it again and let Neovim crash and one should be generated normally.

**Reply by davidcortesortuno** [reply]
It is enabled, I guess the log file should be somewhere. I Googled about the noice.log file but no one reports about it

**Reply by dpetka2001** [reply]
Does your Neovim crash now that it's enabled? Maybe it crashes too quickly and not able to generate a log file because the plugin doesn't have time to load? I don't know what to tell you. For me I can see the log in the path mentioned.

**Reply by davidcortesortuno** [reply]
It seems that the log file is not generated as you say, because it crashes immediately. With the minimal config file the log is not present either. Curiously, I tried disabling `noice` from my neovim setup following https://www.lazyvim.org/configuration/plugins , and it crashes as well. With the minimal setup it works fine :O

**Reply by dpetka2001** [reply]
@davidcortesortuno Plz see [this](https://github.com/LazyVim/LazyVim/discussions/1322#discussioncomment-6743348) comment. The location is different from what I've told you with the minimal repro init file.

**Reply by davidcortesortuno** [reply]
Ok. I just also noticed that the problem seems to be another `ui` plugin, because disabling `noice` still crashes `nvim` when I set one of the options from LazyVim menu. When I disable all `plugins.ui`, I can set the options (like line numbers or spell) with no problem

**Reply by dpetka2001** [reply]
Maybe try enabling them one by one to see which one is responsible for crashing.

**Reply by abeldekat** [reply]
Are you still referring to the minimal_init? 
Can you supply the output of :checkhealth?  That should be possible using the minimal_init, with noice disabled.

**Reply by davidcortesortuno** [reply]
Yes, I'm using the minimal config shown before. The checkhealth command does not show anything wrong

```

==============================================================================
lazy: require("lazy.health").check()

lazy.nvim ~
- OK Git installed
- OK no existing packages found by other package managers
- OK packer_compiled.lua not found
- WARNING {flash.nvim}: unknown key <vscode>

==============================================================================
lazyvim: require("lazyvim.health").check()

LazyVim ~
- OK Using Neovim >= 0.8.0
- OK `git` is installed
- WARNING `rg` is not installed
- WARNING `{ "fd", "fdfind" }` is not installed
- WARNING `lazygit` is not installed

==============================================================================
nvim: require("nvim.health").check()

Configuration ~
- WARNING Missing user config file: /home/david/git/dotfiles/nvim/.repro//config/nvim/init.vim
  - ADVICE:
    - :help |init.vim|

Runtime ~
- OK $VIMRUNTIME: /usr/share/nvim/runtime

Performance ~
- OK Build type: RelWithDebInfo

Remote Plugins ~
- OK Up to date

terminal ~
- key_backspace (kbs) terminfo entry: `key_backspace=^H`
- key_dc (kdch1) terminfo entry: `key_dc=\E[3~`
- $TERM_PROGRAM="tmux"
- $COLORTERM="truecolor"

==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- OK `tree-sitter` found 0.20.6 (ccd6bf554d922596ce905730d98a77af368bba5c) (parser generator, only needed for :TSInstallFromGrammar)
- OK `node` found v20.3.1 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `/home/david/miniconda3/bin/x86_64-conda-linux-gnu-cc` executable found. Selected from { "/home/david/miniconda3/bin/x86_64-conda-linux-gnu-cc", "cc", "gcc", "clang", "cl", "zig" }
  Version: x86_64-conda-linux-gnu-cc (conda-forge gcc 12.1.0-16) 12.1.0
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.4.8-1-default",
  sysname = "Linux",
  version = "#1 SMP PREEMPT_DYNAMIC Thu Aug  3 13:32:15 UTC 2023 (c24ac79)"
} ~

Parser/Features         H L F I J
  - bash                ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - jsdoc               ✓ . . . .
  - json                ✓ ✓ ✓ ✓ .
  - lua                 ✓ ✓ ✓ ✓ ✓
  - luadoc              ✓ . . . .
  - luap                ✓ . . . .
  - markdown            ✓ . ✓ ✓ ✓
  - markdown_inline     ✓ . . . ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - regex               ✓ . . . .
  - tsx                 ✓ ✓ ✓ ✓ ✓
  - typescript          ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vimdoc              ✓ . . . ✓
  - yaml                ✓ ✓ ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~

==============================================================================
provider: health#provider#check

Clipboard (optional) ~
- OK Clipboard tool found: xsel

Python 3 provider (optional) ~
- `g:python3_host_prog` is not set.  Searching for python3 in the environment.
- Multiple python3 executables found.  Set `g:python3_host_prog` to avoid surprises.
- Executable: /home/david/miniconda3/bin/python3
- Other python executable: /usr/bin/python3
- Other python executable: /bin/python3
- Python version: 3.8.10
- pynvim version: 0.4.3
- OK Latest pynvim is installed.

Python virtualenv ~
- OK no $VIRTUAL_ENV

Ruby provider (optional) ~
- Ruby: ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux-gnu]
- WARNING `neovim-ruby-host` not found.
  - ADVICE:
    - Run `gem install neovim` to ensure the neovim RubyGem is installed.
    - Run `gem environment` to ensure the gem bin directory is in $PATH.
    - If you are using rvm/rbenv/chruby, try "rehashing".
    - See :help |g:ruby_host_prog| for non-standard gem installations.
    - You may disable this provider (and warning) by adding `let g:loaded_ruby_provider = 0` to your init.vim

Node.js provider (optional) ~
- Node.js: v20.3.1
- Nvim node.js host: /usr/local/lib/node_modules/neovim/bin/cli.js
- OK Latest "neovim" npm/yarn/pnpm package is installed: 4.10.1

Perl provider (optional) ~
- WARNING "Neovim::Ext" cpan module is not installed
  - ADVICE:
    - See :help |provider-perl| for more information.
    - You may disable this provider (and warning) by adding `let g:loaded_perl_provider = 0` to your init.vim

==============================================================================
vim.lsp: require("vim.lsp.health").check()

- LSP log level : WARN
- Log path: /home/david/git/dotfiles/nvim/.repro//state/nvim/lsp.log
- Log size: 0 KB

vim.lsp: Active Clients ~
- No active clients

==============================================================================
vim.treesitter: require("vim.treesitter.health").check()

- Nvim runtime ABI version: 14
- OK Parser: bash       ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/bash.so
- OK Parser: c          ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/c.so
- OK Parser: html       ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/html.so
- OK Parser: javascript ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/javascript.so
- OK Parser: jsdoc      ABI: 13, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/jsdoc.so
- OK Parser: json       ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/json.so
- OK Parser: lua        ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/lua.so
- OK Parser: luadoc     ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/luadoc.so
- OK Parser: luap       ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/luap.so
- OK Parser: markdown   ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/markdown.so
- OK Parser: markdown_inline ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/markdown_inline.so
- OK Parser: python     ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/python.so
- OK Parser: query      ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/query.so
- OK Parser: regex      ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/regex.so
- OK Parser: tsx        ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/tsx.so
- OK Parser: typescript ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/typescript.so
- OK Parser: vim        ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/vim.so
- OK Parser: vimdoc     ABI: 14, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/vimdoc.so
- OK Parser: yaml       ABI: 13, path: /home/david/git/dotfiles/nvim/.repro/plugins/nvim-treesitter/parser/yaml.so

==============================================================================
which-key: require("which-key.health").check()

WhichKey: checking conflicting keymaps ~
- WARNING conflicting keymap exists for mode **"n"**, lhs: **"gc"**
- rhs: ` `

```

**Reply by abeldekat** [reply]
I also don't see anything special in the health output.

> Ok. I just also noticed that the problem seems to be another ui plugin, because disabling noice still crashes nvim when I set one of the options from LazyVim menu. When I disable all plugins.ui, I can set the options (like line numbers or spell) with no problem

That's very strange. You type "leader u l" and it crashes, only when plugins.ui is enabled? 

Could you disable noice, as you did before, and also add the following:
```lua
{"rcarriga/nvim-notify", enabled = false},
```

**Reply by dpetka2001** [reply]
> Configuration ~
> - WARNING Missing user config file: /home/david/git/dotfiles/nvim/.repro//config/nvim/init.vim

Do you install Lazyvim in `~/.config/nvim` or some other folder? From the minimal it seems like you're putting the configuration in `/home/david/git/dotfiles/nvim/`. But the minimal is configured to look for `XDG_*` directories in the `.repro` folder inside where you put the `minimal_init.lua` file. I don't believe the same can be said for Lazyvim which expects to find standard `XDG_*` directories. @abeldekat do you think that might be a problem or should behave just fine?

**Reply by abeldekat** [reply]
@dpetka2001, I missed that one. It should work I think, if ~/.config/nvim is properly linked to the dotfiles repo. All other plugins work, so the directory mechanism is in place.

I proposed to disable nvim-notify, because "leader u l" only issues a very simple command, and triggers a notification.

**Reply by abeldekat** [reply]
@dpetka2001,

... But it's worth trying out without the dotfiles, on a regular ~/.config/nvim dir.

**Reply by abeldekat** [reply]
@dpetka2001, 
That works on my system, using ln -sf

**Reply by davidcortesortuno** [reply]
Yeah sorry, I forgot to mention I linked the config folder to the git backup. 
If I disable `noice` and also add `{"rcarriga/nvim-notify", enabled = false},` , the `ui` commands no longer crash the session and I can edit the files normally :D 

Now, when I uncomment the `plugins_step2` after disabling the previous 2 plugins, `nvim` crashes again

**Reply by dpetka2001** [reply]
Do you have another Linux distro installed or easily available to try? It might have to do with some runtime specific in Opensuse (in this case a compilation of Neovim from source with debug symbols and an appropriate stacktrace with an issue in Neovim repo might be more helpful). Or spin up a docker with a Ubuntu minimal and set up Neovim and Lazyvim and see how it behaves. Also if you build Neovim from source it might give you a hint if it's not even compiled successfully. Really, Lazyvim is just a plugin in the end and pretty thoroughly tested and shouldn't make Neovim crash. If Neovim crashes, that usually suggests there's something else lying under the hood. I've seen other plugins causing problems that usually show up in messages or fill a screen full of errors, but not make Neovim itself crash (at least from my limited experience and of course I might be wrong).

**Reply by davidcortesortuno** [reply]
Ok, yes, I'll do that. I was thinking to do something similar as the problem became too complex. It is odd, as neovim was working perfectly a few weeks ago, before I updated my distro. The Appimages also have the same problem in my installation.
Anyway, thanks a lot for your time and the help

**Reply by abeldekat** [reply]
@dpetka2001 , I was just about to write a very similar comment. I completely agree, thanks!
@davidcortesortuno, thanks for this discussion. If you find the solution, an update would be appreciated.

**Reply by davidcortesortuno** [reply]
Just to close this issue, I solved the problem by deleting the `.cargo` directory in my `home`. Some weeks ago I installed `neovide` (which did not worked very well) and, I'm not sure why, `nvim` started to crash in any of my terminals. I don't know how the `rust` temporary folder affects `neovim` and `lazyvim` , but it was causing the problem

**Reply by abeldekat** [reply]
Very strange! I am glad the problem is solved. Thanks for the update!

**Comment by abeldekat** [comment]
> You can find the logs for this minimal_init in ~/.config/nvim/.repro/state/nvim

It should be present, when running with noice enabled.

**Reply by abeldekat** [reply]
@dpetka2001 
Thanks! Happy to help.

**Reply by dpetka2001** [reply]
@abeldekat Any particular reason in the import ordering in `plugin_step1`? Just trying to understand the reasoning if there is one indeed.

**Reply by abeldekat** [reply]
Yes, trial and error...)
`Coding.lua` crahses without `treesitter.lua`. 
`Lazy.nvim` operates alphabetically on a single `import = lazyvim.plugins`.

**Reply by amaanq** [reply]
Might be a dup of this
https://github.com/nvim-treesitter/nvim-treesitter/issues/5195
